packages/nimble-components/src/modal/index.ts

```diff
@@ -0,0 +1,135 @@
+import { attr } from '@microsoft/fast-element';
+import { FoundationElement } from '@microsoft/fast-foundation';
+import { UserDismissed } from '../patterns/dialog/types';
+import { ModalState } from './types';
+
+/**
+ * This is a workaround for an incomplete definition of the native dialog element. Remove when using Typescript >=4.8.3.
+ * https://github.com/microsoft/TypeScript/issues/48267
+ * @internal
+ */
+export interface ExtendedDialog extends HTMLDialogElement {
+    showModal(): void;
+    close(): void;
+}
+
+/**
+ * A base class for components based on a native dialog.
+ * @internal
+ */
+// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
+export abstract class Modal<CloseReason = void> extends FoundationElement {
+    /**
+     * @public
+     * @description
+     * Prevents dismissing via the Escape key
+     */
+    @attr({ attribute: 'prevent-dismiss', mode: 'boolean' })
+    public preventDismiss = false;
+
+    /**
+     * The ref to the internal dialog element.
+     *
+     * @internal
+     */
+    public readonly dialogElement!: ExtendedDialog;
+
+    protected get state(): ModalState {
+        return this._state;
+    }
+
+    private set state(value: ModalState) {
+        this._state = value;
+    }
+
+    /**
+     * False if the Modal is closed, true otherwise
+     */
+    public get open(): boolean {
+        return this.state !== ModalState.closed;
+    }
+
+    private _state: ModalState = ModalState.closed;
+    private resolveShow?: (reason: CloseReason | UserDismissed) => void;
+
+    /**
+     * Opens the Modal
+     * @returns Promise that is resolved when the Modal is closed. The value of the resolved Promise is the reason value passed to the close() method, or UserDismissed if closed via the ESC key.
+     */
+    public async show(): Promise<CloseReason | UserDismissed> {
+        if (
+            this.state === ModalState.open
+            || this.state === ModalState.opening
+        ) {
+            throw new Error('Already open or opening');
+        }
+        this.state = ModalState.opening;
+        this.startOpening();
+        return new Promise((resolve, _reject) => {
+            this.resolveShow = resolve;
+        });
+    }
+
+    /**
+     * Closes the Modal
+     * @param reason An optional value indicating how/why the Modal was closed.
+     */
+    public close(reason: CloseReason): void {
+        if (this.state !== ModalState.open) {
+            throw new Error('Not open');
+        }
+        this.state = ModalState.closing;
+        this.startClosing(reason);
+    }
+
+    /**
```

Comment:
The responsibilities of cancel and close handlers don't seem quite right. I think similar to some previous commits they seem over coupled to specific workflows. Modal instances like drawer calling "super.cancelHandler" to start the transition to Closing and check the preventDismiss state seems like unclear responsibilities for a method called "cancelHandler".

Looking at it again I also have a more general question of why are we not designing animations into the Modal base class. I'm actually wondering now if it is really intentional that the dialog has no animations. Seems like even short, fast animations would be useful for all modals.

Some directions:
- We can continue to discuss and iterate on the responsibilities of the base class, including seeing if animation responsibilities make more sense there, it probably makes sense to discuss it first to prevent review cycles or continue the approach of aync comment and feedback if you want to keep going that way
- We pause the refactor for now since we don't have a clear need for shared Modal base classes at the moment (I'm not aware of new modals we need to implement or if we want to share more behavior like animations across all the modals)

Reply:
I'll de-prioritize and move to draft until I have new changes to review.